### DISCUSSION TITLE
What does 'NVO' command do?

### DISCUSSION THREAD

**Discussion by firemeteorxx** [ipxe/ipxe#1291] [Q&A]
I find the following from the config/general.h file.
`
#define NVO_CMD                 /* Non-volatile option storage commands */
`
The comment looks interesting but I can not find any documentation nor any discussion about it.
It appears to be enabled by default but does not work in the shell command line...

I was imagining some kind of persistent state that would allow to save info like the menu choice used in the last time.
Is it something ever possible? (e.g. through UEFI variables in efi environment)

**Comment by NiKiZe** [comment]
https://github.com/ipxe/ipxe/blob/72316b820d4bdbf3d75a0ae7e13f1c3bc8e6ac29/src/hci/commands/nvo_cmd.c basic variables/settings management cmds

**Comment by mcb30** [comment]
It controls the `set`, `show`, etc commands.

The reason for the "NVO" naming is that these commands were originally implemented to handle non-volatile option storage.  There are a few drivers that do provide a non-volatile options settings block, and any settings stored in these will persist across a reboot.  Search for `register_nvo` in the codebase to find which drivers actually support this.

There was never much usage of non-volatile option storage, not least because there is usually very little storage space available.  For example. if you have a Realtek 8139-based card then you have 64 bytes available.  It's just about enough to store a static IPv4 network configuration.

The settings mechanism and its `set`. `show`, `inc`, etc commands turned out to be much more useful for handling general configuration, providing script variables, and for allowing access to a lot of system data (e.g. you can use them to read any value from any SMBIOS table).

There is read-only access to UEFI variables, but no write access.  This is because UEFI variables have an irritating concept of a GUID as well as a variable name, and that doesn't fit neatly into our settings abstraction.  (On read, we return any variable where the name matches, and ignore the GUID.)
